File: Algorithms/0307_Range_Sum_Query_-_Mutable/Python/Range_Sum_Query_-_Mutable_Solution_1.py

# Space: O(n)
# Time: O(n)

class NumArray:

    def __init__(self, nums):
        self.sum = sum(nums)
        self.cache = nums
        self.length = len(nums)

    def update(self, index, val):
        self.sum -= self.cache[index]
        self.sum += val
        self.cache[index] = val

    def sumRange(self, left, right):
        res = self.sum
        if left != 0:
            index = 0
            while index != left:
                res -= self.cache[index]
                index += 1
        if right != self.length - 1:
            index = self.length - 1
            while index != right:
                res -= self.cache[index]
                index -= 1
        return res

File: Algorithms/0307_Range_Sum_Query_-_Mutable/Java/RangeSumQuery-MutableSolution1.java

// Space: O(n)
// Time: O(n)

class NumArray {
    private int[] cache = null;
    private int sum = 0;

    public NumArray(int[] nums) {
        cache = nums;
        sum = Arrays.stream(nums).sum();
    }

    public void update(int index, int val) {
        sum -= cache[index];
        sum += val;
        cache[index] = val;
    }

    public int sumRange(int left, int right) {
        int res = sum;
        int index = 0;
        while (index != left) {
            res -= cache[index];
            index++;
        }
        index = cache.length - 1;
        while (index != right) {
            res -= cache[index];
            index--;
        }
        return res;
    }
}





